File: org.wolfgang/org.wolfgang.contrail-js/src/main/js/Contrail/component/multi/MultiDestinationComponent.js
/*
 * Copyright (C)2012 D. Plaindoux.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation; either version 2, or (at your option) any
 * later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 */

/*global define*/

define([ "require", "Core/jObj", "./flow/MultiUpStreamDataFlow" ],
    function (require, jObj, upStreamDataFlow) {
        "use strict";

        function MultiDestinationComponent() {
            var Factory = require("Component/Factory");
            jObj.bless(this, Factory.core.source(), Factory.core.destinationWithSingleSource());
            this.destinationLink = [];
            this.upStreamDataFlow = upStreamDataFlow(this);
        }

        /**
         * Construction initialisation
         */
        MultiDestinationComponent.init = jObj.constructor([],
            function () {
                return new MultiDestinationComponent();
            });

        MultiDestinationComponent.prototype.acceptDestination = jObj.method([jObj.types.String], jObj.types.Boolean,
            function (componentId) {
                var result = true;

                this.destinationLink.forEach(function (link) {
                    result = result && link.getDestination().getComponentId() !== componentId;
                });

                return result;
            });

        MultiDestinationComponent.prototype.connectDestination = jObj.method([jObj.types.Named("DestinationLink")], jObj.types.Named("ComponentLink"),
            function (destinationLink) {
                this.destinationLink = this.destinationLink.concat(destinationLink);
                return require("Link/Factory").components(this, destinationLink.getDestination());
            });

        MultiDestinationComponent.prototype.getDestinations = jObj.method([], jObj.types.Array,
            function () {
                return this.destinationLink.map(function (link) {
                    return link.getDestination();
                });
            });

        MultiDestinationComponent.prototype.getUpStreamDataFlow = jObj.method([], jObj.types.Named("DataFlow"),
            function () {
                return this.upStreamDataFlow;
            });

        MultiDestinationComponent.prototype.closeUpStream = jObj.procedure([],
            function () {
                this.upStreamDataFlow.handleClose();
                this.destinationLink.forEach(function (link) {
                    link.dispose();
                });
                this.destinationLink = [];
            });

        return MultiDestinationComponent.init;
    });

File: org.wolfgang/org.wolfgang.contrail-core/src/main/java/org/wolfgang/contrail/component/multi/MultiDestinationComponent.java
/*
 * Copyright (C)2012 D. Plaindoux.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation; either version 2, or (at your option) any
 * later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 */

package org.wolfgang.contrail.component.multi;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

import org.wolfgang.contrail.component.ComponentConnectedException;
import org.wolfgang.contrail.component.ComponentConnectionRejectedException;
import org.wolfgang.contrail.component.ComponentDisconnectionRejectedException;
import org.wolfgang.contrail.component.ComponentId;
import org.wolfgang.contrail.component.ComponentNotConnectedException;
import org.wolfgang.contrail.component.SourceComponent;
import org.wolfgang.contrail.component.core.DestinationComponentWithSingleSource;
import org.wolfgang.contrail.component.multi.flow.MultiUpStreamDataFlow;
import org.wolfgang.contrail.flow.DataFlow;
import org.wolfgang.contrail.flow.exception.DataFlowCloseException;
import org.wolfgang.contrail.link.DestinationComponentLink;
import org.wolfgang.contrail.link.DisposableLink;

/**
 * <code>MultiDestinationComponent</code>
 * 
 * @author Didier Plaindoux
 * @version 1.0
 */
public class MultiDestinationComponent<U, D> extends DestinationComponentWithSingleSource<U, D> implements SourceComponent<U, D> {

	private final Map<ComponentId, DestinationComponentLink<U, D>> destinations;
	private final DataFlow<U> upStreamDataFlow;

	{
		this.destinations = new HashMap<ComponentId, DestinationComponentLink<U, D>>();
		this.upStreamDataFlow = new MultiUpStreamDataFlow<U>(this);
	}

	@Override
	public DataFlow<U> getUpStreamDataFlow() {
		return this.upStreamDataFlow;
	}

	public Collection<DestinationComponentLink<U, D>> getDestinationComponentLinks() {
		return destinations.values();
	}

	@Override
	public boolean acceptDestination(ComponentId componentId) {
		return !this.destinations.containsKey(componentId);
	}

	@Override
	public DisposableLink connectDestination(DestinationComponentLink<U, D> handler) throws ComponentConnectionRejectedException {
		final ComponentId componentId = handler.getDestinationComponent().getComponentId();
		if (this.acceptDestination(componentId)) {
			this.destinations.put(componentId, handler);
			return new DisposableLink() {
				@Override
				public void dispose() throws ComponentDisconnectionRejectedException {
					if (destinations.remove(componentId) == null) {
						throw new ComponentNotConnectedException(NOT_YET_CONNECTED.format());
					}
				}
			};

		} else {
			throw new ComponentConnectedException(ALREADY_CONNECTED.format());
		}
	}

	@Override
	public void closeUpStream() throws DataFlowCloseException {
		for (DestinationComponentLink<U, D> link : getDestinationComponentLinks()) {
			link.getDestinationComponent().closeUpStream();
		}
	}
}
